### PR TITLE
ddl: disallow changing the default value of the primary key column to null

### LINKED ISSUE #24265: tidb could change parts of primary to null
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
CREATE TABLE `PK_MULTI_COL_50` (  `COL1` varchar(10) NOT NULL,  `COL2` binary(1) NOT NULL,  PRIMARY KEY (`COL1`(5),`COL2`) /*T![clustered_index] CLUSTERED */) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
alter table PK_MULTI_COL_50 modify col1 varchar(30) null;

```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
1. should not modify the primary key type of cluster index
2. should not be set part of primary to null
```sql
mysql >  alter table PK_MULTI_COL_50_tmp modify col1 varchar(30) null;
(1171, 'All parts of a PRIMARY KEY must be NOT NULL; if you need NULL in a key, use UNIQUE instead')
```
### 3. What did you see instead (Required)
```sql
mysql > alter table PK_MULTI_COL_50 modify col1 varchar(30) null;
Query OK, 0 rows affected
Time: 6.690s
mysql > show create table  PK_MULTI_COL_50;
+-----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table           | Create Table                                                                                                                                                                                                                   |
+-----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| PK_MULTI_COL_50 | CREATE TABLE `PK_MULTI_COL_50` (\n  `col1` varchar(30) DEFAULT NULL,\n  `COL2` binary(1) NOT NULL,\n  PRIMARY KEY (`col1`(5),`COL2`) /*T![clustered_index] CLUSTERED */\n) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
```mysql
tidb_version() | Release Version: v4.0.0-beta.2-2707-g6889ab44a-dirty
Edition: Community
Git Commit Hash: 6889ab44a5bf21ec0465cd84a4baf68072f1bd3c
Git Branch: master
UTC Build Time: 2021-04-22 13:12:44
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

```

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close #24265 

Problem Summary:
Handle the issue of changing the default value of the primary key column to null.

### What is changed and how it works?

What's Changed:

- Add `checkPriKeyConstraint` for `modify/change column`.
- Uniform processing and checking of default values and column properties.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

### Release note 

- Disallow the default value of the primary key column to be NULL when using the 'modify/change column' operation.

### CODE REVIEW

File: ddl/column_type_change_test.go [hunk 1]
```diff
@@ -1702,6 +1702,24 @@ func (s *testColumnTypeChangeSuite) TestChangingAttributeOfColumnWithFK(c *C) {
 	tk.MustExec("drop table if exists orders, users")
 }
 
+func (s *testColumnTypeChangeSuite) TestAlterPrimaryKeyToNull(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	// Enable column change variable.
+	tk.Se.GetSessionVars().EnableChangeColumnType = true
+	defer func() {
```
Comment (wjhuang2016) [inline ddl/column_type_change_test.go:1710]: Defer is not needed since it's a session variable?

File: ddl/column_type_change_test.go [hunk 2]
```diff
@@ -1702,6 +1702,21 @@ func (s *testColumnTypeChangeSuite) TestChangingAttributeOfColumnWithFK(c *C) {
 	tk.MustExec("drop table if exists orders, users")
 }
 
+func (s *testColumnTypeChangeSuite) TestAlterPrimaryKeyToNull(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	// Enable column change variable.
```
Comment (tangenta) [inline ddl/column_type_change_test.go:1708]:
```suggestion
```

**Review by tangenta — APPROVED** [review]

**Pull request merged**